src/Workspaces/Remote/Core/Services/RemotePasteTrackingService.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PasteTracking;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    [Export(typeof(IPasteTrackingService))]
+    [Shared]
+    internal sealed class RemotePasteTrackingService : IPasteTrackingService
+    {
+        private readonly Dictionary<SourceTextContainer, (TextSpan? span, int referenceCount)> _trackedSpans = new Dictionary<SourceTextContainer, (TextSpan? span, int referenceCount)>();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RemotePasteTrackingService()
+        {
+        }
+
+        public bool TryGetPastedTextSpan(SourceTextContainer sourceTextContainer, out TextSpan textSpan)
+        {
+            lock (_trackedSpans)
+            {
+                var result = _trackedSpans.TryGetValue(sourceTextContainer, out var spanAndCount);
+                if (!result || spanAndCount.referenceCount == 0 || spanAndCount.span is null)
+                {
+                    textSpan = default;
+                    return false;
+                }
+
+                textSpan = spanAndCount.span.Value;
+                return result;
+            }
+        }
+
+        private void AddPastedTextSpanReference(SourceTextContainer container, TextSpan? pastedTextSpan)
+        {
+            lock (_trackedSpans)
+            {
+                if (!_trackedSpans.TryGetValue(container, out var spanAndCount))
+                {
+                    _trackedSpans[container] = (pastedTextSpan, 1);
+                    return;
+                }
+
+                if (spanAndCount.span != pastedTextSpan)
+                {
+                    // Cannot track two different pasted spans for the same container
+                    throw ExceptionUtilities.UnexpectedValue(pastedTextSpan);
```

Comment:
@CyrusNajmabadi The other option here is to assign `null` (no pasted text span) if multiple callers cannot agree on the same non-null span. As it stands, code that attempts to use a different span will fail if prior work has not yet completed.